packages/@aws-cdk/core/test/arn.test.ts

```diff
@@ -175,10 +191,32 @@ nodeunitShim({
         },
       };
 
-      Object.keys(tests).forEach(arn => {
-        const expected = tests[arn];
-        test.deepEqual(stack.parseArn(arn), expected, arn);
-      });
+      for (const [arn, expectedComponents] of Object.entries(tests)) {
+        const skipCheckingCfnEncoding = expectedComponents.checkCfnEncoding === false;
+        // delete the extra field so it doesn't screw up the equality comparison
+        delete expectedComponents.checkCfnEncoding;
+
+        // test the basic case
+        const parsedComponents = stack.parseArn(arn);
+        test.deepEqual(parsedComponents, expectedComponents, arn);
+
+        // test the round-trip
+        test.equal(stack.formatArn(parsedComponents), arn);
+
+        // test that the CloudFormation functions we generate evaluate to the correct value
+        if (skipCheckingCfnEncoding) {
+          continue;
+        }
+        const tokenArnComponents = stack.parseArn(
+          Token.asString(new Intrinsic({ Ref: 'TheArn' })),
+          parsedComponents.sep,
+          parsedComponents.resourceName !== undefined,
+          parsedComponents.resourcePrefixed);
+        const formattedTokenArn = stack.formatArn(tokenArnComponents);
```

Comment:
Mmmm. I *guess* this works? :)

Struggling to get my head around if this extra operation has the potential to hide bugs... but I suppose not?

Reply:
Not sure I follow what you're saying here.

What's the alternative way to formulate this test without `formatArn()`?

Reply:
Testing the evaluation of the components (which I did in my sample code -- does it not work?)

Reply:
OK, changed.